File: c++/design_pattern/Decorator.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <initializer_list>

using namespace std;


namespace Decorator{
	
	struct Notifier {
		virtual void send(const string& message) =0;
	};

	struct ConcreteNotifier : Notifier {
		void send(const string& message){
			cout << "[ConcreteNotifier] : " << message << endl;
		}
	};

	struct BaseDecorator : Notifier{
		
		BaseDecorator(Notifier& wrapee) : wrapee(wrapee) {}

		void send(const string& message) override{
			wrapee.send(message);
		}

		protected : 
			Notifier& wrapee;
	};

	class SMSDecorator : public BaseDecorator {
		public : 
			
			SMSDecorator(Notifier& wrapee) : BaseDecorator(wrapee){}

			void sendSMS(const string& message){
				cout << "[SMS] :" << message << endl;
			}

			void send(const string& message) override{
				BaseDecorator::send(message);
				sendSMS(message);
			}
	};

	class SlackDecorator : public BaseDecorator {
		public : 
			
			SlackDecorator(Notifier& wrapee) : BaseDecorator(wrapee){}

			void sendSlack(const string& message){
				cout << "[Slack] :" << message << endl;
			}

			void send(const string& message) {
				BaseDecorator::send(message);
				sendSlack(message);
			}
	};
	class FacebookDecorator : public BaseDecorator {
		public : 
			
			FacebookDecorator(Notifier& wrapee) : BaseDecorator(wrapee){}

			void sendFacebook(const string& message){
				cout << "[Facebook] :" << message << endl;
			}

			void send(const string& message) {
				BaseDecorator::send(message);
				sendFacebook(message);
			}
	};

	void main(){

		ConcreteNotifier cn;
		SlackDecorator sd(cn);
		FacebookDecorator fd(sd);
		SMSDecorator smsd(fd);

		smsd.send("hello world");

	}
};

struct Shape {
	virtual string str() const = 0;
};

struct Circle : Shape {
	
	float radius;
	explicit Circle(const float radius) : radius(radius) {}
	void resize(float factor){ radius *= factor;}
	string str() const override {
		ostringstream oss;
		oss << "A circle of radius " << radius;
		return oss.str();
	}

};

namespace DynamicDecortator {

	struct ColorShape : Shape {
		Shape& shape;
		string color; 

		ColorShape(Shape& shape, const string& color) : shape(shape), color(color){}
		ColorShape(Shape&& shape, const string& color) : shape(shape), color(color){}

		string str() const override{
			ostringstream oss;
			oss << shape.str() << " has " << color;
			return oss.str();
		}

	};

	struct TransparentShape : Shape {
		Shape& shape;
		uint8_t transparency; 

		TransparentShape(Shape& shape, const uint8_t transparency) : shape(shape), transparency(transparency){}
		TransparentShape(Shape&& shape, const uint8_t transparency) : shape(shape), transparency(transparency){}

		string str() const override{
			ostringstream oss;
			oss << shape.str() << " has " << static_cast<float>(transparency)/255.f * 100.f << " transparency";
			return oss.str();
		}
	};
	
	void main(){

		cout << "[good Case]"<<endl ;
		TransparentShape c1{
			ColorShape{
				Circle(23), "red"
			}, 128
		};
		cout << c1.str() << endl;


		cout << "[Bad Case]"<<endl ;
		ColorShape c2{
			ColorShape{
				Circle(23), "red"
			}, "green"
		};
		cout << c2.str() << endl;

	}
};

namespace StaticDecorator {


	template <typename T>
	struct ColorShape : T{

		string color;

		template <typename ...Args>
		ColorShape(const string& color, Args...args)
			: T(std::forward<Args>(args)...), color(color){}

		string str() const override {
			ostringstream oss;
			oss << T::str() << " has the color " << color ;
			return oss.str();
		}

	};


	template <typename T>
	struct TransparentShape : T {

		uint8_t transparency;

		template <typename ...Args>
		TransparentShape(uint8_t transparency, Args... args) 
			: T(std::forward<Args>(args)...), transparency(transparency){}

		string str() const override {
			ostringstream oss;
			oss << T::str() << " has the transparency " << transparency / 255.f * 100.f;
			return oss.str();
		}

	};

	void main(){
	
		ColorShape<TransparentShape<Circle>> c3 {"red", 64, 5};
		cout << c3.str() << endl;
	
	}

};

namespace FunctionalDecorator {

	// 1. function 타입이용

	struct Logger {
		function<void()> func;
		string name; 

		Logger(const function<void()>& func, const string& name) 
			: func{func}, name{name} {}

		void operator()() const {
			cout << "Engering " << name << endl;
			func();
			cout << "Exiting " << name << endl;
		}
	};

	// 2. 템플릿 인자 이용

	template<typename Func>
	struct TemplateLogger{
		string name;
		Func func;

		TemplateLogger(const Func& func, const string& name)
			: func{func}, name{name} {}

		void operator()() const {
			cout << "Engering " << name << endl;
			func();
			cout << "Exiting " << name << endl;
		}		
	};

	template <typename Func> auto make_logger ( Func func, const string& name){
		return TemplateLogger<Func>{func, name};
	}	
	
	void main(){

		// 1. function 타입 이용
		cout << "=========" << endl;
		Logger([](){
			cout << "Hello" << endl;
		}, "HelloFunction")();


		// 2. 템플릿 인자 이용
		cout << "=========" << endl;
		auto call = make_logger([](){ cout << "Hello"<<endl;}, "HelloFunction2");
		call();
	}
};


int main(void){

	Decorator::main();
	DynamicDecortator::main();
	StaticDecorator::main();
	FunctionalDecorator::main();

	return 0;
}

File: java/src/main/java/com/example/design_pattern/Decorator.java
package com.example.design_pattern;

import java.util.*;
import java.lang.reflect.*;
import java.util.function.*;

public class Decorator{
	
	private static class Decorator_{


		private static interface Notifier {
			void send(String message);
		}

		private static class ConcreteNotifier implements Notifier {
			@Override
			public void send(String message){
				System.out.println("[Concrete] : " + message);
			}
		}

		private static class BaseDecorator  implements Notifier{
			protected Notifier wrapee;

			BaseDecorator(Notifier wrapee){
				this.wrapee = wrapee;
			}

			@Override
			public void send(String message){
				wrapee.send(message);
			}
		}

		private static class SMSDecorator extends BaseDecorator {
			SMSDecorator(Notifier wrapee){
				super(wrapee);
			}

			private void sendSMS(String message){
				System.out.println("[SMS] : " + message);
			}

			@Override
			public void send(String message){
				super.send(message);
				sendSMS(message);
			}
		}

		private static class SlackDecorator extends BaseDecorator {
			SlackDecorator(Notifier wrapee){
				super(wrapee);
			}

			private void sendSlack(String message){
				System.out.println("[Slack] : "+message);
			}

			@Override
			public void send(String message){
				super.send(message);
				sendSlack(message);
			}
		}

		public static void main(){
			SMSDecorator decoratedNotifier = new SMSDecorator(
				new SlackDecorator(
					new ConcreteNotifier()
				));

			decoratedNotifier.send("Hello World");
		}
	}


	private static class DyncamicDecorator {

		private static interface Shape {
			String render();
		}

		private static class Circle implements Shape {
			double radius;

			public Circle(double radius){
				this.radius = (double) radius;
			}

			public void resize(double factor){
				radius *= factor;
			}

			@Override
			public String render(){
				return String.format("Circle of raidus %.2f", radius);
			}

		}


		private static class ColorShape implements Shape {
			private Shape shape;
			private String color;

			ColorShape(Shape shape, String color){
				this.shape = shape;
				this.color = color;
			}

			@Override
			public String render(){
				return String.format("%s has %s", shape.render(), color);
			}

		}

		private static class TransparentShape implements Shape {
			private Shape shape;
			private int transparency;

			public TransparentShape(Shape shape, int transparency){
				this.shape = shape;
				this.transparency = transparency;
			}

			@Override
			public String render(){
				return String.format("%s has %.2f%% transparency", shape.render(), ((double) Math.min(256,transparency) * 100) /256 );
			}
		}

		public static void main(){


			ColorShape decoratedShape = new ColorShape(
				new TransparentShape(
					new Circle(3.0), 128
				),
				"red"
			);

			System.out.println(decoratedShape.render());

		}
	}

	private static class StaticDecorator_ {

		private static interface Shape {
			String render();
		}

		private static class Circle implements Shape {
			double radius;

			public Circle(Object... radius){
				if(radius[0] instanceof Number){
					this.radius = (double) radius[0];
					return;
				}
				this.radius = 0;
			}

			public void resize(double factor){
				radius *= factor;
			}

			@Override
			public String render(){
				return String.format("Circle of raidus %.2f", radius);
			}

		}

		private static class StaticDecorator <T extends Shape> implements Shape {
			
			protected T shape;
			protected static Object[] trimObjects(Object... args){
				return Arrays.stream(args)
							.skip(1)
							.toArray();
			}

			public StaticDecorator(Object... args) {

				try {
					Class<T> clazz = (Class<T>) args[0];
					Object[] arguments = {};
					if(args.length > 1){
						arguments = trimObjects(args);	

					} 

					Class<?>[] classes = new Class<?>[arguments.length];
					for(int i=0; i<arguments.length; i++){
						classes[i] = arguments[i].getClass();
					}

					Constructor<?> constructor = clazz.getConstructors()[0];
					shape = (T) constructor.newInstance((Object) arguments);

				}catch(InstantiationException | IllegalAccessException 
					| InvocationTargetException e){
					
					throw new IllegalArgumentException(e);
				}
			}

			@Override
			public String render(){
				return shape.render();
			}
		}


		private static class ColorShape<T extends Shape> extends StaticDecorator {

			private String color;

			public ColorShape(Object... args){
				super(StaticDecorator.trimObjects(args));
				this.color = (String) args[0];
			}

			@Override
			public String render(){
				return String.format("%s has %s", super.render(), color );
			}

		}

		private static class TransparentShape<T extends Shape> extends StaticDecorator {

			private int transparency;

			public TransparentShape(Object... args){
				super(StaticDecorator.trimObjects(args));
				this.transparency = (int) args[0];
			}

			@Override
			public String render(){
				return String.format("%s has %.2f%% transparency", super.render(), ((double) Math.min(transparency,256)*100) / 256  );
			}
		}


		public static void main(){

			ColorShape<TransparentShape<Circle>> decoratedShape = 
				new ColorShape("red", TransparentShape.class, 128, Circle.class, 3.0);

			System.out.println(decoratedShape.render());
		}
	}

	private static class FunctionalDecorator {


		private static class Logger implements Runnable{
			
			private Runnable function;
			private String name;

			Logger(String name, Runnable function){
				this.name = name;
				this.function = function;
			}

			@Override
			public void run(){
				System.out.println("Entering " + name);
				function.run();
				System.out.println("Exiting " + name);
			}
		}

		public static void main(){
			Logger logger = new  Logger("Hello Function",()->{
				System.out.println("Hello");
			});

			logger.run();
		}
	}


	public static void main(String... args){

		Decorator_.main();
		DyncamicDecorator.main();
		StaticDecorator_.main();
		FunctionalDecorator.main();
	}
}